File: src/fairseq2/native/extension/data/text/converters.cc
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

#include "fairseq2/native/extension/module.h"

#include <cstdint>
#include <optional>
#include <stdexcept>
#include <string_view>
#include <utility>
#include <vector>

#include <ATen/ScalarType.h>

#include <fairseq2/native/data/data_pipeline.h>
#include <fairseq2/native/data/text/string_splitter.h>
#include <fairseq2/native/data/text/string_to_int_converter.h>
#include <fairseq2/native/data/text/string_to_tensor_converter.h>

namespace py = pybind11;

namespace fairseq2 {

void
def_text_converters(py::module_ &text_module)
{
    py::module_ m = text_module.def_submodule("converters");

    // StrSplitter
    py::class_<string_splitter, std::shared_ptr<string_splitter>>(m, "StrSplitter")
        .def(
            py::init([](std::string_view sep, std::optional<std::vector<std::string>> names)
            {
                if (sep.size() != 1)
                    throw std::invalid_argument{"`sep` must be of length 1."};

                return string_splitter{sep[0], std::move(names)};
            }),
            py::arg("sep") = '\t',
            py::arg("names") = std::nullopt)
        .def("__call__", &string_splitter::operator(), py::call_guard<py::gil_scoped_release>{});

    // StrToIntConverter
    py::class_<
        string_to_int_converter, std::shared_ptr<string_to_int_converter>>(m, "StrToIntConverter")
        .def(py::init<std::int16_t>(), py::arg("base") = 10)
        .def("__call__", &string_to_int_converter::operator());

    // StrToTensorConverter
    py::class_<
        string_to_tensor_converter,
        std::shared_ptr<string_to_tensor_converter>>(m, "StrToTensorConverter")
        .def(
            py::init<std::optional<std::vector<std::int64_t>>, std::optional<at::ScalarType>>(),
            py::arg("size") = std::nullopt,
            py::arg("dtype") = std::nullopt)
        .def(
            "__call__",
            &string_to_tensor_converter::operator(),
            py::call_guard<py::gil_scoped_release>{});

    map_functors().register_<string_splitter>();
    map_functors().register_<string_to_int_converter>();
    map_functors().register_<string_to_tensor_converter>();
}

}  // namespace fairseq2

File: src/fairseq2/data/text/converters.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

from __future__ import annotations

from collections.abc import Sequence
from typing import TYPE_CHECKING, final

from fairseq2n import DOC_MODE
from torch import Tensor

from fairseq2.data_type import DataType

if TYPE_CHECKING or DOC_MODE:

    @final
    class StrSplitter:
        """Split string on a given character.

        :param sep:
            The character to split on (default to tab)

        :param names:
            names of the corresponding columns of the input tsv file
            Will create dictionaries object with one entry per column

        :param indices:
            A list of indices of the column to keep, or a single index.
            If a single index is provided and ``exclude`` is ``False``,
            the output is a string.

        :param exclude:
            If ``True``, the indices will be excluded from the output,
            instead of kept. Default to ``False``.

        Example usage::

            # read all columns: ["Go.", "Va !", "CC-BY 2.0 (France)"]
            dataloader = read_text("tatoeba.tsv").map(StrSplitter()).and_return()
            # keep first and second columns, yielding the list: ["Go.", "Va !"]
            dataloader = read_text("tatoeba.tsv").map(StrSplitter(indices=[0, 1])).and_return()
            # keep only the second column, directly yielding a string: "Va !"
            dataloader = read_text("tatoeba.tsv").map(StrSplitter(indices=1)).and_return()
            # keep only the first and second column and convert to dict: {"en": "Go.", "fr": "Va !"}
            dataloader = read_text("tatoeba.tsv").map(StrSplitter(names=["en", "fr"], indices=[0, 1])).and_return()

        """

        def __init__(
            self,
            sep: str = "\t",
            names: Sequence[str] | None = None,
            indices: int | Sequence[int] | None = None,
            exclude: bool = False,
        ) -> None: ...

        def __call__(self, s: str) -> str | list[str] | dict[str, str]: ...

    @final
    class StrToIntConverter:
        """Parses integers in a given base"""

        def __init__(self, base: int = 10) -> None: ...

        def __call__(self, s: str) -> int: ...

    @final
    class StrToTensorConverter:
        def __init__(
            self,
            size: Sequence[int] | None = None,
            dtype: DataType | None = None,
        ) -> None: ...

        def __call__(self, s: str) -> Tensor: ...

else:
    from fairseq2n.bindings.data.text.converters import (  # noqa: F401
        StrSplitter as StrSplitter,
    )
    from fairseq2n.bindings.data.text.converters import (  # noqa: F401
        StrToIntConverter as StrToIntConverter,
    )
    from fairseq2n.bindings.data.text.converters import (  # noqa: F401
        StrToTensorConverter as StrToTensorConverter,
    )
